### PR TITLE
Supporting ETag http headers on static files

### PR DESCRIPTION
**Pull request by mathertel**
The ETag (or entity tag) HTTP header specified the version of the returned resource.
This is done by the default implementation by calculating the md5 checksuom of the file content.
A custom ETag calculation can be passed as an option.

This allows client side caching to be more efficient and save bandwidth.

The implementation is taken from esp8266/Arduino and is compatible with the WebServer implementation for esp8266.
where it was implemented by @imwhocodes and me (@mathertel).

The `WebServer` now include the methon `enableETag(bool enable, ETagFunction fn = nullptr)`
to enable and disable ETag Headers for a server.

The `StaticRequestHandler` calculates the ETag value when enabled and checks / populates the header.

An optional ETag calculation function can be passed to enableETag (e.g. to provide different algorithms).

## Example

The **WebServer Example** illustrates using the ETag function.

It is using the "last change date" of the filesystem to create ETags wich is more efficient.

It is available on some file systems implementations (LittleFS, FSFAT but not SPIFFS)
and requires NTP to provide a current datetime.

### about http ETags

ETags are "hints" delivered with a web response in the `ETag` http-Header to the client.

This hint value is added to further requests to the same resource in the .`If-None-Match` Header.
Hereby the server can check if the client should still use the version in the cache or needs an update.

The web server does not send a full response and file content when the content was not changed
but just returns a `304 Not Modified` status to the client resulting in much smaller network packages.

In the cases when the clients will request for the same resources multiple times in a row
the server-side checksum check is a lot more effective than transferring the bytes over the network.

As this is not always true there is an extended option to inject a function that behavior can be customized and tailored to the specific requirements.

### enabling ETag support

To enable this in the embedded web server the enableETag() can be used.
(next to enableCORS)

In the simplest version just call enableETag(true) to enable the internal ETag generation that calcs the hint using a md5 checksum in base64 encoded form. This is an simple approach that adds some time for calculation on every request but avoids network traffic.

The headers will look like:

``` txt
If-None-Match: "GhZka3HevoaEBbtQOgOqlA=="
ETag: "GhZka3HevoaEBbtQOgOqlA=="
```

### ETag support customization

The enableETag() function has an optional second optional parameter to provide a function for ETag calculation of files.

The function enables eTags for all files by using calculating a value from the last write timestamp:

``` cpp
server.enableETag(true, [](FS &fs, const String &path) -> String {
  File f = fs.open(path, "r");
  String eTag = String(f.getLastWrite(), 16);  // use file modification timestamp to create ETag
  f.close();
  return (eTag);
});
```

The headers will look like:

``` txt
ETag: "63bbaeb5"
If-None-Match: "63bbaeb5"
```

### CODE REVIEW

**Comment by VojtechBartoska** [comment]
Thanks for nice and complete PR @mathertel! We'll review it soon.

**Review by PilnyTomas — CHANGES_REQUESTED** [review]
@mathertel nice work!
Would it be easy to add some extended feedback on the file upload? For example in serial output mention the file name and size and/or enable downloading it back.
Also please modify the `.ino` header and `README.md` so that it is using our [new template](https://github.com/espressif/arduino-esp32/tree/563804706b0ad76b6eae433e342f1d8790dacfcd/libraries/ESP32/examples/Template/ExampleTemplate)
Otherwise LGTM.

**Comment by P-R-O-C-H-Y** [comment]
Any updates @mathertel on @PilnyTomas 's comment?

**Comment by mathertel** [comment]
please review.

**Comment by PilnyTomas** [comment]
@mathertel Thanks for the addition. Please make sure that when someone tries to upload a too large file it doesn't run away with a looped message:
```
./components/esp_littlefs/src/littlefs/lfs.c:584:error: No more free space 385
./components/esp_littlefs/src/littlefs/lfs.c:584:error: No more free space 385
./components/esp_littlefs/src/littlefs/lfs.c:584:error: No more free space 385
And to the infinity ...
```
For small files, the messages are enough.

**Comment by mathertel** [comment]
@PilnyTomas : All implemented as you suggested. Please Review or give some feedback.

**Comment by mathertel** [comment]
This is a compatibility request to ESP8266 to easy migration to ESP32 (C3).
Would be welcome to include it in the next version.
@PilnyTomas : All implemented as you suggested. Please Review or give some feedback.
@P-R-O-C-H-Y : Who can do the review when @PilnyTomas doesn't find the time ?

**Comment by VojtechBartoska** [comment]
hello @mathertel, thanks for the updates. We'll review your PR soon.

**Review by PilnyTomas — APPROVED** [review]
Hi, sorry for the delay, I had some happy changes in my life, so I took a few weeks off.
Thank you for your updates.
I would love to see a different message when the upload fails - instead of "finished" there could be something like "Upload failed"
It isn't necessary - just nice to have, so I approve the PR.

**Review by P-R-O-C-H-Y — CHANGES_REQUESTED** [review]
@mathertel Please add a empty `.skip.esp32h2` file to your example folder to fix failing CI.

**Comment by mathertel** [comment]
> @mathertel Please add a empty `.skip.esp32h2` file to your example folder to fix failing CI.

done.

**Review by lucasssvaz — APPROVED** [review]
Small nitpick but LGTM

File: libraries/WebServer/examples/WebServer/data/files.htm [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          fetch(fname, { method: 'DELETE' });
+          document.location.reload(false);
+        }
+      };
+    });
+
+  </script>
+</body>
+
+</html>
```
[the remark this answers is not in the document]
Comment (me-no-dev) [reply libraries/WebServer/examples/WebServer/data/files.htm:65]: I think you fixed `index.htm` (which was also to be fixed), but forgot to also add a line at the end in `files.htm` as well

**Pull request merged**